### PR TITLE
use runtime-wasm feature instead of assuming no_std to be wasm

### LINKED ISSUE #5547: Don't assume wasm in no_std env in runtime-interface
Currently [sp-io](https://github.com/paritytech/substrate/blob/master/primitives/io/src/lib.rs) implements external functions for the runtime. It uses `runtime-interface` macros to automatically generate the functions for native and wasm build.

Specifically, it is handled by:
- [host_function_interface](https://github.com/paritytech/substrate/blob/master/primitives/runtime-interface/proc-macro/src/runtime_interface/host_function_interface.rs)
- [bare_function_interface](https://github.com/paritytech/substrate/blob/master/primitives/runtime-interface/proc-macro/src/runtime_interface/host_function_interface.rs)

As written in the code, the macros determines if the generated interface function is for wasm or native code by checking `std` feature.

It usually works, until someone wants use the runtime **natively** in a `no_std` environment. `runtime-interface` will try to generate a wasm external call when it detects a `no_std` environment, but there's no wasm interface because it's actually a native build. As a result, it will trigger a link error.

The problem lies in the way `runtime-interface` detects if it's a wasm or native build. I suggest to add a new rust feature as a switch (e.g. `runtime_wasm`), instead of checking std or no_std. With that I can simply build the runtime in a native no_std target.

_Why native no_std?_

I'm working on [a Substrate light client in TEE](https://github.com/Phala-Network/phala-pruntime/tree/master/enclave/src/light_validation). The TEE enclave code is in no_std because it's required by the TEE SDK.

In the light client, there are a few places calculating the hash of the block header (`header.hash()`), which eventually calls `sp_io::hashing::blake2_256()`. The hash functions in `sp_io` is defined by `runtime-interface`. So if any code invokes `block.hash()`, it will try to find the wasm interface and thus fails to link.

Currently my walk-around is to replace `header.hash()` by serializing the header and call `sp_core::hashing::blake2_256()` directly, as written in [wasm_hacks](https://github.com/Phala-Network/phala-pruntime/blob/master/enclave/src/light_validation/wasm_hacks.rs). If this issue can be fixed, it will be much easier to make the light client a separate package for better code reuse. I believe this cloud also help substraTEE team.

### PR DESCRIPTION
**Pull request by brenzi**
fixes #5547 

Main steps done:
1. replace `#[cfg(feature = "std")]` by `#[cfg(not(feature = "runtime-wasm"))]` and the logical opposite
```
find . -type f -name "*.rs" ! -path "./target/*" -print0 | xargs -0 sed -i 's/#\[cfg(feature = "std")\]/#\[cfg(not(feature = "runtime-wasm"))\]/g' 
find . -type f -name "*.rs" ! -path "./target/*" -print0 | xargs -0 sed -i 's/#\[cfg(not(feature = "std"))\]/#\[cfg(feature = "runtime-wasm")\]/g' 
```
2. introduce `runtime-wasm` feature in all tomls (probably that's too many as not all are needed for runtime)
```
find . -type f -name "*.toml" -print0 | xargs -0 sed -i 's/default = \[ "std" \]/default = \[ "std" \]\nruntime-wasm = \[\]/g' 
find . -type f -name "*.toml" -print0 | xargs -0 sed -i 's/default = \["std"\]/default = \[ "std" \]\nruntime-wasm = \[\]/g' 
```
3. let wasm-builder-runner add the `runtime-wasm` feature to rustc args
4. manual fixes, i.e. to gate `RuntimeDebug` derivation

this is WIP. not yet sure if the approach makes sense

### CODE REVIEW

File: bin/node-template/runtime/build.rs [hunk 1]
```diff
@@ -6,5 +6,6 @@ fn main() {
 		.with_wasm_builder_from_crates("2.0.0")
 		.export_heap_base()
 		.import_memory()
+		.append_to_rust_flags("--cfg feature=\\\"runtime-wasm\\\"")
```
Comment (bkchr) [inline bin/node-template/runtime/build.rs:9]: Why?

File: bin/node/runtime/build.rs [hunk 2]
```diff
@@ -23,5 +23,6 @@ fn main() {
 		.with_wasm_builder_from_crates_or_path("2.0.0", "../../../utils/wasm-builder")
 		.export_heap_base()
 		.import_memory()
+		.append_to_rust_flags("--cfg feature=\\\"runtime-wasm\\\"")
```
Comment (bkchr) [inline bin/node/runtime/build.rs:26]: ?

**Review by bkchr — CHANGES_REQUESTED** [review]
In general the `runtime-wasm` feature should only be used where it is required!

This means mostly `primitives-io`, because most other crates are still using the `no_std` "legally" by making themselves compilable for no-std.

Everywhere where we have special code that assumes `no_std` == `runtime-wasm`, we would need to switch it to `runtime-wasm`.

File: bin/node-template/runtime/build.rs (hunk 1 above)
Comment (brenzi) [reply bin/node-template/runtime/build.rs:9]: from the errors I concluded that the feature `runtime-wasm` isn't applied automatically by the build script. Is that wrong?

**Comment by brenzi** [comment]
I just tried to reset and try bottom up starting with `sp-io` only. However, the errors explode and I think it takes someone with deep knowledge of the code to distinguish between `no_std == wasm` assumptions and legit `std` switches.

Try and err will lead us nowhere because iterations are long as everything is rebuilt upon changes to fundamental crates.

**Comment by brenzi** [comment]
I would assume we need to touch at least theses:
* `sp-io`
* `sp-debug-derive`
* `runtime-interface`

TBC

File: bin/node-template/runtime/build.rs (hunk 1 above)
Comment (bkchr) [reply bin/node-template/runtime/build.rs:9]: When the feature is present in the cargo.toml, the wasm-builder should activate it automatically.

**Comment by bkchr** [comment]
> I just tried to reset and try bottom up starting with `sp-io` only. However, the errors explode and I think it takes someone with deep knowledge of the code to distinguish between `no_std == wasm` assumptions and legit `std` switches.
> 
> Try and err will lead us nowhere because iterations are long as everything is rebuilt upon changes to fundamental crates.

I assume you desperately waiting for this feature? :see_no_evil:

**Comment by brenzi** [comment]
well, it would make our live a lot easier for sure :pray:

**Comment by brenzi** [comment]
well, I tried a little harder, but giving up for today....

**Comment by athei** [comment]
Does this kind of "negative feature" even work with `Cargo.toml`? How do you express that you want to enable the `std` feature of some (non parity) dependency when `runtime-wasm` is **not** set?

**Comment by gnunicorn** [comment]
Closing for a lack of progress.

**Pull request closed without merging**